File: apps/hash-graph/lib/graph/src/api/rest/utoipa_typedef/subgraph/vertices/vertex.rs
use serde::{Deserialize, Serialize};
use type_system::url::BaseUrl;
use utoipa::ToSchema;

use crate::{
    identifier::ontology::OntologyTypeVersion,
    knowledge::Entity,
    ontology::{DataTypeWithMetadata, EntityTypeWithMetadata, PropertyTypeWithMetadata},
    subgraph::identifier::{DataTypeVertexId, EntityTypeVertexId, PropertyTypeVertexId},
};

// WARNING: will always deserialize as `DataType`
#[derive(Debug, Clone, Hash, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
#[serde(untagged)]
#[expect(clippy::enum_variant_names)]
pub enum OntologyTypeVertexId {
    DataType(DataTypeVertexId),
    PropertyType(PropertyTypeVertexId),
    EntityType(EntityTypeVertexId),
}

impl OntologyTypeVertexId {
    #[must_use]
    pub const fn base_id(&self) -> &BaseUrl {
        match self {
            Self::DataType(id) => &id.base_id,
            Self::PropertyType(id) => &id.base_id,
            Self::EntityType(id) => &id.base_id,
        }
    }

    #[must_use]
    pub const fn revision_id(&self) -> OntologyTypeVersion {
        match self {
            Self::DataType(id) => id.revision_id,
            Self::PropertyType(id) => id.revision_id,
            Self::EntityType(id) => id.revision_id,
        }
    }
}

#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
#[serde(tag = "kind", content = "inner")]
#[serde(rename_all = "camelCase")]
#[expect(clippy::enum_variant_names)]
pub enum OntologyVertex {
    #[schema(title = "DataTypeVertex")]
    DataType(Box<DataTypeWithMetadata>),
    #[schema(title = "PropertyTypeVertex")]
    PropertyType(Box<PropertyTypeWithMetadata>),
    #[schema(title = "EntityTypeVertex")]
    EntityType(Box<EntityTypeWithMetadata>),
}

impl From<DataTypeWithMetadata> for OntologyVertex {
    fn from(data_type: DataTypeWithMetadata) -> Self {
        Self::DataType(Box::new(data_type))
    }
}

impl From<PropertyTypeWithMetadata> for OntologyVertex {
    fn from(property_type: PropertyTypeWithMetadata) -> Self {
        Self::PropertyType(Box::new(property_type))
    }
}

impl From<EntityTypeWithMetadata> for OntologyVertex {
    fn from(entity_type: EntityTypeWithMetadata) -> Self {
        Self::EntityType(Box::new(entity_type))
    }
}

#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
#[serde(tag = "kind", content = "inner")]
#[serde(rename_all = "camelCase")]
pub enum KnowledgeGraphVertex {
    #[schema(title = "EntityVertex")]
    Entity(Entity),
}

#[derive(Debug, PartialEq, Eq, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
#[serde(untagged)]
#[expect(dead_code, reason = "This is used in the generated OpenAPI spec")]
pub enum Vertex {
    #[schema(title = "OntologyVertex")]
    Ontology(Box<OntologyVertex>),
    #[schema(title = "KnowledgeGraphVertex")]
    KnowledgeGraph(Box<KnowledgeGraphVertex>),
}

File: libs/@local/hash-subgraph/src/types/vertex.ts
import { BaseUri } from "@blockprotocol/type-system";

import {
  DataTypeWithMetadata,
  Entity,
  EntityTypeWithMetadata,
  PropertyTypeWithMetadata,
} from "./element";
import { EntityId, Timestamp } from "./shared";

// -------------------------------- Vertex Variants --------------------------------

export type DataTypeVertex = { kind: "dataType"; inner: DataTypeWithMetadata };

export type PropertyTypeVertex = {
  kind: "propertyType";
  inner: PropertyTypeWithMetadata;
};

export type EntityTypeVertex = {
  kind: "entityType";
  inner: EntityTypeWithMetadata;
};

export type EntityVertex = { kind: "entity"; inner: Entity };

export type OntologyVertex =
  | DataTypeVertex
  | PropertyTypeVertex
  | EntityTypeVertex;

export type KnowledgeGraphVertex = EntityVertex;

export type Vertex = OntologyVertex | KnowledgeGraphVertex;

export const isDataTypeVertex = (vertex: Vertex): vertex is DataTypeVertex => {
  return vertex.kind === "dataType";
};

export const isPropertyTypeVertex = (
  vertex: Vertex,
): vertex is PropertyTypeVertex => {
  return vertex.kind === "propertyType";
};

export const isEntityTypeVertex = (
  vertex: Vertex,
): vertex is EntityTypeVertex => {
  return vertex.kind === "entityType";
};

export const isEntityVertex = (vertex: Vertex): vertex is EntityVertex => {
  return vertex.kind === "entity";
};

// -------------------------------- The `Vertices` type --------------------------------

export type OntologyVertices = {
  [typeBaseUri: BaseUri]: {
    [typeVersion: number]: OntologyVertex;
  };
};

export type KnowledgeGraphVertices = {
  [entityId: EntityId]: {
    [entityVersion: Timestamp]: KnowledgeGraphVertex;
  };
};

export type Vertices = OntologyVertices & KnowledgeGraphVertices;
